u_tls_extensions.go

```diff
@@ -242,6 +242,7 @@ type ALPNExtension struct {
 func (e *ALPNExtension) writeToUConn(uc *UConn) error {
 	uc.config.NextProtos = e.AlpnProtocols
 	uc.HandshakeState.Hello.AlpnProtocols = e.AlpnProtocols
+	uc.HandshakeState.Hello.NextProtoNeg = uc.HandshakeState.Hello.NextProtoNeg || len(e.AlpnProtocols) > 0
```

Comment:
Ah yes I should have mentioned this to you - I was meaning to ask you about it.

Currently, without that line, the `TestUTLSFingerprintClientHello` test fails when checking mimicry against the `Randomized-ALPN` spec:
```
NextProtoNeg fields not equal, expected: true, got: false
```

I _believe_ that's because when creating the randomized spec, it actually touches the underlying `UConn`'s `Config`:
https://github.com/refraction-networking/utls/blob/4ee624457953e3730be8cb012f0cc0450867d1c9/u_parrots.go#L810-L817

This in turn means that when `makeClientHello` is called, the hello's `NextProtoNeg` is set true:
https://github.com/refraction-networking/utls/blob/master/handshake_client.go#L77-L79

However, `FingerprintClientHello` doesn't actually have access to a `Config`, so it can't make the kind of parallel change.

I'm honestly struggling a tiny bit with fully understanding the interaction between `ALPN` and `NPN` re: the golang crypto implementation, but I _think_ that this makes sense. I'm *very* open to an explanation of how I'm wrong.

Reply:
Default Golang behavior is that if config.NextProtos is not nil, it will enable both NPN and ALPN. I think it's important that we set it correctly because we may run into the issues in this chunk of code:
https://github.com/refraction-networking/utls/blob/43c36d3c1f57546d5cbb05c066df7b5a78686c51/handshake_client.go#L681-L699

The correct fix would be to look if our fingerprint actually has NPN by creating a bool `haveNPN` and then filling it when we are cycling through the extensions here:
https://github.com/refraction-networking/utls/blob/33a29038e742910d6ec82636748dfad5dd7f30ae/u_parrots.go#L621
and then `config.NextProtoNeg =  haveNPN`.
Please leave a comment why we're doing this: default makeClientHello always sets NextProtoNeg and sends the NPN extension if NextProtos, also used by ALPN, is set, but we may not actually have NPN in our ClientHelloSpec.

Reply:
Great ok took me a little while to wrap my head around this, but I believe I understand what you're saying and I'll make the fix.